hw/ip/otbn/dv/rig/rig/gens/edge_store.py

```diff
@@ -0,0 +1,93 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import Optional
+
+from shared.operand import ImmOperandType, RegOperandType
+from shared.insn_yaml import InsnsFile
+
+from ..config import Config
+from ..model import Model
+from ..program import ProgInsn, Program
+
+from ..snippet import ProgSnippet
+from ..snippet_gen import GenCont, GenRet, SnippetGen
+
+
+class EdgeStore(SnippetGen):
+    '''A snippet generator that generates random stores that tries to
+    hit edge cases.
+
+    '''
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        self.sw = self._get_named_insn(insns_file, 'sw')
+
+        # SW has three operands: grs1, offset, grs2
+        if not (len(self.sw.operands) == 3 and
+                isinstance(self.sw.operands[0].op_type, RegOperandType) and
+                self.sw.operands[0].op_type.reg_type == 'gpr' and
+                not self.sw.operands[0].op_type.is_dest() and
+                isinstance(self.sw.operands[1].op_type, ImmOperandType) and
+                isinstance(self.sw.operands[2].op_type, RegOperandType) and
+                self.sw.operands[2].op_type.reg_type == 'gpr' and
+                not self.sw.operands[2].op_type.is_dest()):
+            raise RuntimeError('SW instruction from instructions file is '
+                               'not the shape expected by the EdgeLoadStore generator.')
+
+        self.grs1_op_type = self.sw.operands[0].op_type
+
+        self.imm_op_type = self.sw.operands[1].op_type
+        assert self.imm_op_type.shift == 0
+
+        self.grs2_op_type = self.sw.operands[2].op_type
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 1:
+            return None
+
+        # ImmOperand has a specific width hence offset cannot be None
+        offset_rng = self.imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        base = []
+        known_regs = model.regs_with_known_vals('gpr')
+        # Pick a small negative/near the top positive known value for hitting edges
+        for reg_idx, reg_val in known_regs:
+            if (6140 >= reg_val >= 2050):  # If reg_val reach very close to the top
```

Comment:
I assume these numbers are "top of dmem +- imm range"? Could you compute them explicitly?